README.md
# luno-node

Use the Luno API in Node.js

This module is a very thin wrapper on the Luno API, handling authentication, request signing and errors.

## Install

```sh
npm install --save luno
```

## Usage

For documentation on all methods and routes, please see our [docs](https://luno.io/docs).

```js
var Luno = require('luno');
var luno = new Luno({
  key: 'YOUR-API-KEY', // Your Luno API key
  secret: 'YOUR-SECRET-KEY', // Your Luno secret key
  timeout: 10000, // Maximum request timeout (in milliseconds). Default 10000.
  sandbox: false // Set to true to enable Sandbox Mode for all requests (unless otherwise specified in params).  See https://luno.io/docs#sandbox
});

luno.get('/users/usr_cz2D0VOugcBVWW', {}, function(err, user) {

});
```

### Methods

GET, POST, PUT, PATCH, DELETE and request.

```js
// GET
// luno.get(route, query, callback);
luno.get('/users', {
  limit: 10
}, function(err, resp) { });

// POST
// luno.post(route, query, body, callback);
luno.post('/sessions', {}, {
  user_id: 'usr_cz2D0VOugcBVWW',
  ip: req.headers['cf-connecting-ip'] || req.headers['x-forwarded-for'] || req.connection.remoteAddress,
  user_agent: req.headers['user-agent'],
  details: {
    custom: 'session',
    properties: 'here'
  }
}, function(err, session) { });

// PUT
// luno.put(route, query, body, callback);
luno.put('/users/usr_cz2D0VOugcBVWW', {}, {
  name: 'Simon Tabor',
  profile: {
    properties: 'are',
    overwritten: true
  }
}, function(err, resp) { });

// PATCH
// luno.patch(route, query, body, callback);
luno.patch('/users/usr_cz2D0VOugcBVWW', {}, {
  name: 'Simon Tabor',
  profile: {
    properties: 'are',
    extended: true,
    old: 'properties',
    are: 'kept'
  }
}, function(err, resp) { });

// DELETE
// luno.delete(route, query, callback);
luno.delete('/users/usr_cz2D0VOugcBVWW', {}, function(err, resp) { });


// REQUEST (any method)
// luno.request(METHOD, route, query, body, callback);
luno.request('POST', '/users', {}, {
  email: 'simon@luno.io',
  password: 'my-password'
}, function(err, user) { });
```

### Middleware

Use the session middleware to quickly ensure a session is valid and fetch the details.

It'll set `req.session` to the session details (if the session is valid) and `req.user` to the user details (if the session is valid and there is an associated user).

```js
// app is an express server

app.use(luno.session({
  cookieName: 'session', // default config
  cookieConfig: {
    maxAge: 1209600000, // 14 days
    httpOnly: true,
    // secure: true
  }
}));

// admin section of the app requires the session to be valid
// and to have an associated user
app.use('/admin', function(req, res, next) {
  if (!req.user) return res.redirect('/login');
  next();
});

app.get('/', function(req, res, next) {
  // req.session and req.user might be falsy

  if (req.user) {
    res.send('Welcome ' + req.user.first_name);
  } else {
    res.send('Welcome');
  }
});

app.get('/admin', function(req, res, next) {
  // req.session and req.user must be set

  res.send('Hello ' + req.user.first_name);
});
```

### Registration

Let a user sign up and log them in.

*Note:* you may want to look at [Luno CSRF](https://github.com/lunoio/luno-csrf-node) to add CSRF protection using Luno.

```js
app.get('/signup', function(req, res) {
  // Send a barebones signup form
  res.send('<form method="POST" action="/signup"><input type="email" name="email"><input type="password" name="password"><input type="submit"></form>');
});

app.post('/signup', function(req, res, next) {
  luno.post('/users', {}, {
    email: req.body.email,
    password: req.body.password,
    profile: {
      test: true
    }
  }, function(err, user) {
    if (err) return next(err);

    // automatically log this user in
    luno.post('/sessions', {}, {
      user_id: user.id,
      ip: req.connection.remoteAddress,
      user_agent: req.headers['user-agent']
    }, function(err, session) {
      if (err) return next(err);

      res.cookie('session', session.key, {
        maxAge: 1209600000, // 14 days
        httpOnly: true,
        secure: true
      });

      // redirect to the login-protected application
      res.redirect('/dashboard');
    });
  });
});
```


CHANGELOG.md
# luno-node changelog

## [1.1.0](https://github.com/lunoio/luno-node/tree/v1.1.0) (2016-04-21)
[Compare](https://github.com/lunoio/luno-node/compare/v1.0.5...v1.1.0)

- Added JSDoc comments to main functions
- Added argument validation to ensure proper usage and highlight mistakes
- Added utils to handle module errors and contain common functions in the future

## [1.0.5](https://github.com/lunoio/luno-node/tree/v1.0.5) (2016-04-06)
[Compare](https://github.com/lunoio/luno-node/compare/v1.0.4...v1.0.5)

- Sandbox Mode can now be configured for all requests and overridden per-request, rather than having to be specified on every request.

## [1.0.4](https://github.com/lunoio/luno-node/tree/v1.0.4) (2016-03-30)
[Compare](https://github.com/lunoio/luno-node/compare/v1.0.3...v1.0.4)

- Routes can now include `/v1` prefix without resulting in a `route_not_found` error ([\#1](https://github.com/lunoio/luno-node/issues/1))

## [1.0.3](https://github.com/lunoio/luno-node/tree/v1.0.3) (2016-03-21)
[Compare](https://github.com/lunoio/luno-node/compare/v1.0.2...v1.0.3)

- Errors now have a `description` property, which is returned by the API to supplement the `message` with more details.

## [1.0.2](https://github.com/lunoio/luno-node/tree/v1.0.2) (2016-02-29)
[Compare](https://github.com/lunoio/luno-node/compare/v1.0.1...v1.0.2)

- Added [Session middleware](https://github.com/lunoio/luno-node/tree/v1.0.2#middleware) to easily run `/sessions/access` code, setting `req.session` with the current active session for a user.

## [1.0.1](https://github.com/lunoio/luno-node/tree/v1.0.1) (2016-02-09)
[Compare](https://github.com/lunoio/luno-node/compare/v1.0.0...v1.0.1)

- Fixed HMAC signatures for requests that include UTF-8 characters (e.g. emoji).

## [1.0.0](https://github.com/lunoio/luno-node/tree/v1.0.0) (2016-02-08)

- Initial release


lib/session.js
module.exports = function(luno, conf) {

  var config = {
    cookieName: 'session',
    cookieConfig: {
      maxAge: 1209600000,
      httpOnly: true
    }
  };

  for (var i in conf) config[i] = conf[i];

  return function(req, res, next) {
    var key = req.cookies[config.cookieName];
    if (!key) return next();

    luno.post('/sessions/access', { expand: 'user' }, {
      key: key,
      ip: req.headers['cf-connecting-ip'] || req.headers['x-forwarded-for'] || req.connection.remoteAddress,
      user_agent: req.headers['user-agent']
    }, function(err, session) {
      if (err) {
        if (err.code === 'session_not_found') res.clearCookie(config.cookieName, config.cookieConfig);
        return next();
      }

      res.cookie(config.cookieName, session.key, config.cookieConfig);

      req.session = session;
      req.user = session.user;
      next();
    });
  };
};


lib/API.js
'use strict';
var crypto = require('crypto');
var request = require('request');
var qs = require('querystring');
var pkg = require('../package.json');
var session = require('./session');
var utils = require('./utils');

/**
 * Luno class
 * @date   2016-04-21
 * @param  {Object}   config Configuration properties to override defaults.
 *                           Key and secret are required, see https://luno.io/dashboard/account
 */
var Luno = module.exports = function(config) {
  var self = this;

  if (typeof config !== 'object' || !config) throw utils.invalidArgument('config', config, 'an object')
  if (typeof config.key !== 'string' || !config.key) throw utils.invalidArgument('config.key', config.key, 'a string');
  if (typeof config.secret !== 'string' || !config.secret) throw utils.invalidArgument('config.secret', config.secret, 'a string');

  self.config = {
    host: 'https://api.luno.io',
    version: 1,
    key: null,
    secret: null,
    timeout: 10000,
    sandbox: false
  };

  for (var i in config) self.config[i] = config[i];

  self.endpoint = '/v' + self.config.version;
};

/**
 * Make a signed request to Luno's API
 * @date   2016-04-21
 * @param  {String}       method HTTP method to use, e.g. GET or POST
 * @param  {String}       route  Endpoint to make the request to, e.g. /users
 * @param  {Object}       params Query-string parameters to send with the request
 * @param  {Object|Null}  body   Entity/POST body to send with the request
 * @param  {Function}     cb     Callback function
 */
Luno.prototype.request = function(method, route, params, body, cb) {
  var self = this;

  if (typeof cb !== 'function') throw utils.invalidArgument('callback', cb, 'a function');
  if (typeof method !== 'string') return cb(utils.invalidArgument('method', method, 'a string'));
  if (typeof route !== 'string') return cb(utils.invalidArgument('route', route, 'a string'));
  if (typeof params !== 'object' || !params) return cb(utils.invalidArgument('params', params, 'an object'));
  if (typeof body !== 'object') return cb(utils.invalidArgument('body', body, 'null or an object'));

  var validMethods = [ 'GET', 'POST', 'PUT', 'PATCH', 'DELETE' ];
  if (validMethods.indexOf(method) === -1) {
    return cb(utils.invalidArgument('method', method, 'in ' + validMethods.join(', ')));
  }

  var p = {};
  for (var i in params) {
    p[i] = params[i];
  }

  p.key = self.config.key;
  p.timestamp = new Date().toISOString();
  if (typeof p.sandbox === 'undefined' && self.config.sandbox) p.sandbox = true;

  if (route.indexOf(self.endpoint) !== 0) route = self.endpoint + route;

  var sign = method + ':' + route + '?' + qs.stringify(p);
  if (body) sign += ':' + JSON.stringify(body);
  p.sign = crypto.createHmac('sha512', self.config.secret).update(sign, 'utf8').digest('hex');

  request({
    url: self.config.host + route,
    method: method,
    qs: p,
    body: body ? body : undefined,
    json: true,
    timeout: self.config.timeout,
    headers: {
      'user-agent': 'luno_node/' + pkg.version
    }
  }, function(err, res, data) {
    if (err) return cb(err);

    if (!data) {
      return cb(utils.error({
        message: 'Empty response',
        code: 'empty_response',
        status: res.statusCode
      }));
    }

    // non-200 statuscode - error
    if (res.statusCode >= 300 || res.statusCode < 200) {
      // not entirely sure what the error is (shouldn't happen)
      if (!data.message) return cb(data);
      return cb(utils.error(data));
    }

    cb(err, data);
  });
};

Luno.prototype.get = function(route, params, cb) {
  this.request('GET', route, params, null, cb);
};

Luno.prototype.post = function(route, params, body, cb) {
  this.request('POST', route, params, body, cb);
};

Luno.prototype.put = function(route, params, body, cb) {
  this.request('PUT', route, params, body, cb);
};

Luno.prototype.patch = function(route, params, body, cb) {
  this.request('PATCH', route, params, body, cb);
};

Luno.prototype.delete = function(route, params, cb) {
  this.request('DELETE', route, params, null, cb);
};

/**
 * Middleware for accessing a session. See https://github.com/lunoio/luno-node#middleware
 * @date   2016-04-21
 * @param  {Object}   config Configuration to override defaults (see lib/session.js for defaults)
 * @return {Function}        Middleware (req, res, next) that checks for a valid session and will access it.
 *                           Sets req.session and req.user to the session and user if it's valid.
 */
Luno.prototype.session = function(config) {
  var self = this;
  return session(self, config);
};
